### ISSUE TITLE
feat: SGD optimizer with momentum support

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#199]
## Summary

The `SGD` optimizer is **vanilla gradient descent** without momentum. Standard ImageNet training uses SGD with momentum 0.9, which is the classic baseline optimizer for ResNets and often outperforms AdamW on CNNs.

## Current State

```java
// SGD.java — no momentum, no state
@Override
public void step(List<ParametricLayer> layers) {
    for (var layer : layers) {
        var params = layer.parameters();
        var grads = layer.gradients();
        for (int i = 0; i < params.size(); i++) {
            computeOps.sgdUpdate(params.get(i), grads.get(i), learningRate);
        }
    }
}

// CpuOps.sgdUpdate — simple param -= lr * grad
void sgdUpdate(Tensor param, Tensor grad, float lr) {
    // SIMD loop: param[i] -= lr * grad[i]
}
```

**Problem**: No velocity/momentum tracking. Each step only uses the current gradient, ignoring gradient history. This causes:
- Noisy convergence (oscillation in narrow valleys)
- Slower convergence on large-scale problems
- Inability to reproduce standard ImageNet baselines (SGD + momentum 0.9 + weight decay 1e-4)

### Updated SGD API

```java
public final class SGD implements Optimizer {
    private final float learningRate;
    private final float momentum;          // NEW (default 0.0 = vanilla SGD)
    private final float weightDecay;       // NEW (default 0.0)
    private final boolean nesterov;        // NEW (default false)
    private final IdentityHashMap<Tensor, float[]> velocities;  // NEW

    // Backward-compatible constructor
    public SGD(float learningRate) {
        this(learningRate, 0f, 0f, false);
    }

    // Full constructor
    public SGD(float learningRate, float momentum, float weightDecay, boolean nesterov) {
        this.learningRate = learningRate;
        this.momentum = momentum;
        this.weightDecay = weightDecay;
        this.nesterov = nesterov;
        this.velocities = new IdentityHashMap<>();
    }
}
```

### SGD with Momentum Algorithm

```
// Standard momentum:
v_t = momentum * v_{t-1} + gradient_t + weightDecay * param
param = param - lr * v_t

// Nesterov momentum (optional):
v_t = momentum * v_{t-1} + gradient_t + weightDecay * param
param = param - lr * (gradient_t + momentum * v_t)
```

### ComputeOps Extension

```java
// New method in ComputeOps interface:
void sgdMomentumUpdate(Tensor param, Tensor grad, float[] velocity,
                        float lr, float momentum, float weightDecay, boolean nesterov);
```

### SIMD Implementation (CpuOps)

```java
void sgdMomentumUpdate(Tensor param, Tensor grad, float[] v,
                        float lr, float mu, float wd, boolean nesterov) {
    float[] p = param.data();
    float[] g = grad.data();

    var vMu = FloatVector.broadcast(SPECIES, mu);
    var vLr = FloatVector.broadcast(SPECIES, lr);
    var vWd = FloatVector.broadcast(SPECIES, wd);

    int i = 0;
    for (; i <= p.length - LANE_COUNT; i += LANE_COUNT) {
        var vg = FloatVector.fromArray(SPECIES, g, i);
        var vp = FloatVector.fromArray(SPECIES, p, i);
        var vv = FloatVector.fromArray(SPECIES, v, i);

        // g_wd = g + wd * p
        var gwd = vg.add(vWd.mul(vp));
        // v = mu * v + g_wd
        vv = vMu.mul(vv).add(gwd);
        vv.intoArray(v, i);

        if (nesterov) {
            // p -= lr * (g_wd + mu * v)
            vp.sub(vLr.mul(gwd.add(vMu.mul(vv)))).intoArray(p, i);
        } else {
            // p -= lr * v
            vp.sub(vLr.mul(vv)).intoArray(p, i);
        }
    }
    // scalar remainder...
}
```

### Usage — Classic ImageNet baseline

```java
var optimizer = new SGD(0.1f, 0.9f, 1e-4f, false);  // lr=0.1, momentum=0.9, wd=1e-4
var scheduler = LrScheduler.stepDecay(0.1f, 30);      // ×0.1 every 30 epochs

var trainer = new Trainer(model, new CrossEntropyLoss(), optimizer)
    .withScheduler(scheduler)
    .withGradientClipping(1.0f);
```

## Acceptance Criteria

- [ ] `SGD(lr)` constructor preserves vanilla behavior (backward compatible)
- [ ] `SGD(lr, momentum, weightDecay, nesterov)` enables full-featured SGD
- [ ] Velocity state tracked per-parameter via `IdentityHashMap`
- [ ] `ComputeOps.sgdMomentumUpdate()` added with SIMD implementation in `CpuOps`
- [ ] Nesterov momentum option implemented correctly
- [ ] Decoupled weight decay (applied to param, not grad)
- [ ] State restore for checkpoint resumption (like AdamW.restoreState)
- [ ] Unit test: momentum accumulates across steps
- [ ] Unit test: weight decay reduces parameter magnitude
- [ ] Unit test: nesterov produces different updates than standard momentum
- [ ] Unit test: vanilla SGD (momentum=0) matches existing behavior exactly

## References

- [PyTorch SGD](https://pytorch.org/docs/stable/generated/torch.optim.SGD.html)
- [On the importance of initialization and momentum in deep learning (Sutskever et al., 2013)](https://proceedings.mlr.press/v28/sutskever13.html)
- [ResNet Paper](https://arxiv.org/abs/1512.03385) — Uses SGD with momentum 0.9, weight decay 1e-4
